### PR TITLE
NCCL Participant should wait on new event, not stream

### PR DESCRIPTION
**Pull request by jeffdaily**
The Collective will not start until all Participant instances arrive.
However, the Participant will call ThenWaitFor on the compute stream
which will wait for all enqueued kernels instead of just the kernel
that is producing the input tensor.  This greatly affects mGPU perf.

### CODE REVIEW

**Comment by jeffdaily** [comment]
Currently labeled as DO NOT MERGE because this should be verified against real data for convergence first.  However, the code review (and CI) should begin immediately.

**Comment by jeffdaily** [comment]
Convergence validation is underway.  Perf results are promising.

# resnet50 using RCCL, without this PR
num_gpus | average | stddev | scaling | efficiency
-- | -- | -- | -- | --
1 | 242.4 | 0.4537253 | 1 | 1
2 | 464.37 | 1.5286813 | 1.915718 | 0.957859
4 | 837.6967 | 4.1623257 | 3.455844 | 0.863961
8 | 1537.24 | 0.3802631 | 6.341749 | 0.792719

# resnet50 using RCCL, with PR
num_gpus | average | stddev | scaling | efficiency
-- | -- | -- | -- | --
1 | 241.06 | 0.791581 | 1 | 1
2 | 480.8033 | 0.744595 | 1.994538 | 0.997269
4 | 924.3633 | 1.414457 | 3.834578 | 0.958644
8 | 1811.763 | 1.442459 | 7.515819 | 0.939477

File: tensorflow/core/nccl/nccl_manager.h [hunk 1]
```diff
@@ -78,6 +79,17 @@ class NcclManager {
       DCHECK(event_mgr != nullptr);
       DCHECK(tensor_stream != nullptr);
       DCHECK(nccl_stream != nullptr);
+      if (input != nullptr) {
+        input_event = new se::Event(executor);
```
Comment (whchung) [inline tensorflow/core/nccl/nccl_manager.h]: @jeffdaily what if we use `unique_ptr`?

Comment (jeffdaily) [reply tensorflow/core/nccl/nccl_manager.h]: Good call.  I will change this to `absl::make_unique` and remove the explicit `Participant` destructor.

Comment (whchung) [reply tensorflow/core/nccl/nccl_manager.h]: @jeffdaily CI test result looks fine. failures in `rocm-xla` should have no relationship to this PR. please revise the code and we can take it.

**Comment by jeffdaily** [comment]
Though I did not run for the full 90 epochs, convergence is evident.  Removing the DO NOT MERGE.

**Pull request merged**